### ISSUE TITLE
bug: some backends don't detect paths with spaces correctly

### ISSUE THREAD

**Issue by Zeioth** [Zeioth/compiler.nvim#43]
A fix is on the way. We just have to make sure we surround all paths by `""`.

**Comment by Zeioth** [comment]
## TODO

- [ ] asm
- [ ] c
- [ ] cpp
- [ ] cs
- [ ] dart
- [ ] elixir
- [ ] fortran
- [ ] fsharp
- [ ] go
- [ ] java
- [ ] javascript
- [ ] kotlin
- [ ] lua
- [ ] make
- [ ] perl
- [ ] python
- [ ] r
- [ ] ruby
- [ ] rust
- [ ] sh
- [ ] swift
- [ ] typescript
- [ ] vb
- [ ] zig

**Comment by Zeioth** [comment]
All tests passing. But this is a big commit. Let's research it for several days.

**Comment by Gaweringo** [comment]
Nice, I've tried it with a Makefile and it worked. But for example for C files this command is used:
```bash
rm -f "/home/gaweringo/compiler test/bin/program" || true && mkdir -p "/home/gaweringo/compiler test/bin/" && gcc /home/gaweringo/compiler test/main.c -o "/home/gaweringo/compiler test/bin/program" -Wall -g && "/home/gaweringo/compiler test/bin/program" && echo "/home/gaweringo/compiler test/main.c" && echo '--task finished--'
```
```bash
.... && gcc /home/gaweringo/compiler test/main.c -o ....
```
Where the argument to `gcc` is not quoted.
Making `utils.find_files` return files passed through `utils.os_path` fixes this:
https://github.com/Gaweringo/compiler.nvim/commit/5a0806d1841316a930fdf1ddd1f6bf7348866925

**Comment by Zeioth** [comment]
@Gaweringo you are right, I missed that. Also, I think I'm gonna add spaces to our tests to ensure this is working for all languages.

**Comment by Gaweringo** [comment]
Probably a good idea. 

I tried to use the testing setup you have, but I had the problem, that the `tests` folder (or the whole `compiler.nvim` folder) was not in `~/.local/share/nvim/lazy/`. I ended up just symlinking the folder there.
Probably didn't set lazy up correctly for loading local/dev plugins.

I'd be interested in working a bit on Windows comparability and having the same test setup as you would likely be good for that.

**Comment by Zeioth** [comment]
Yes sorry for that. For now the tests are Unix only, but I'm open to PRs if anyone want to make the windows compatible. It's really a matter of changing the path to wathever that OS uses.

**Comment by Zeioth** [comment]
Let's give it a round 2 

## TODO (languages)

- [] asm
- [] c
- [] cpp
- [] cs
- [] dart
- [] elixir
- [] fortran
- [] fsharp
- [] go
- [] java
- [] javascript
- [] kotlin
- [] lua
- [] make
- [] perl
- [] python
- [] r
- [] ruby
- [] rust
- [] sh
- [] swift
- [] typescript
- [] vb
- [] zig

## TODO (bau)

- [] cmake
- [] gradle
- [] make
- [] meson
- [] nodejs

**Comment by Zeioth** [comment]
Reverting changes for now as this issue is much more complex than initially expected.

**Comment by Zeioth** [comment]
I think the nuclear option of modifying the functions in utils.lua is not a good approach. We are are operating too many strings. 

Currently testing the next approach:

* Strings are surrounded in the command itself
* In find_files, offer the option to surround.
* In there cases, the backend is reponsible of surrounding every file, in the few cases where they got to be operated (like in asm).
* Adititonally, we have to surround the `executables` of solutions.

## Adventages
* We can visually assert the executed command is correct.
* No weird tricks.
* Even though it makes readability worse, it doesn't add complexity.

**Comment by Zeioth** [comment]
@Gaweringo if you wanna test the branch spaces support before I merge it, all should be fine now. All tests passing with spaces.

**Comment by Gaweringo** [comment]
Just tried it with C, C++, Go, Make and CMake and they all worked. Nice.
